### PR TITLE
Use MurmurHash3 to allow for fast in-memory hashing with no conversion

### PR DESCRIPTION
**Pull request by ScottPJones**

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -1,13 +1,15 @@
 name = "ShortStrings"
 uuid = "63221d1c-8677-4ff0-9126-0ff0817b4975"
 authors = ["Dai ZJ <zhuojia.dai@gmail.com>"]
-version = "0.2.6"
+version = "0.3.0"
```
Comment (oxinabox) [inline Project.toml:4]: I don't think this is breaking?

File: src/base.jl [hunk 2]
```diff
@@ -44,27 +50,30 @@ Base.codeunit(s::ShortString) = UInt8
...
+
+Base.sizeof(s::ShortString{T}) where {T} = Int(s.size_content & (size_mask(s) % UInt))
 Base.firstindex(::ShortString) = 1
 Base.isvalid(s::ShortString, i::Integer) = isvalid(String(s), i)
 Base.iterate(s::ShortString) = iterate(String(s))
 Base.iterate(s::ShortString, i::Integer) = iterate(String(s), i)
 Base.lastindex(s::ShortString) = sizeof(s)
 Base.ncodeunits(s::ShortString) = sizeof(s)
+
+Base.display(s::ShortString) = display(String(s))
```
Comment (oxinabox) [inline src/base.jl:65]: isn't `show` enough?

File: src/hash.jl [hunk 3]
```diff
@@ -1,5 +1,4 @@
-export hash
+using MurmurHash3: mmhash128_a
 
-import Base.hash
-
-Base.hash(x::ShortString, h::UInt) = hash(String(x), h)
+Base.hash(x::ShortString, h::UInt) =
+    last(mmhash128_a(sizeof(x), ntoh(x.size_content), h%UInt32)) + h
```
Comment (oxinabox) [inline src/hash.jl:4]:
```suggestion
    last(mmhash128_a(sizeof(x), bswap(x.size_content), h%UInt32)) + h
```

**Comment by oxinabox** [comment]
Do you have benchmarks?

File: src/hash.jl (hunk 3 above)
Comment (ScottPJones) [reply src/hash.jl]: Sure, what's the difference?

**Comment by ScottPJones** [comment]
I'm not sure that I trust `@benchmark` on this, it is saying about 9.5ns for `String`, about 9ns for `ASCIIStr`, and 0.045ns for `ShortString15`.
I may have to figure some other way to benchmark it.

File: Project.toml (hunk 1 above)
Comment (ScottPJones) [reply Project.toml:4]: You're right!

File: src/base.jl (hunk 2 above)
Comment (ScottPJones) [reply src/base.jl:65]: I just moved that, to organize them a bit better, but I think you are right.

File: src/hash.jl (hunk 3 above)
Comment (xiaodaigh) [reply src/hash.jl]: ntoh is short for network to host, so if host (the computer executing it) is little endian or big little endian it does it differently right? So `bswap` will always swap them? I am bit confused. here.

Comment (oxinabox) [reply src/hash.jl]: `bswap` will always swap.
`nthoh` will not swap if julia is running on a big endian machine.
Only if it is on a little endian machine.
Julia only runs on little endian machines however.
Main advantage is that `bswap` says what it does without remembering  network to host.

**Review by oxinabox — APPROVED** [review]

**Comment by oxinabox** [comment]
> 0.045ns for ShortString15.

Nice, it is constant-folding.
That's great.
ShortStrings constant-fold really well.

File: src/hash.jl [hunk 4]
```diff
@@ -1,5 +1,6 @@
-export hash
+using MurmurHash3: mmhash128_a
 
-import Base.hash
-
-Base.hash(x::ShortString, h::UInt) = hash(String(x), h)
+function Base.hash(x::ShortString, h::UInt)
+    h += Base.memhash_seed
+    last(mmhash128_a(sizeof(x), bswap(x.size_content), h%UInt32)) + h
```
Comment (oxinabox) [inline src/hash.jl:5]: I wonder if this should be:
```suggestion
    last(mmhash128_a(ncodeunits(x), bswap(x.size_content), h%UInt32)) + h
```
to be more semantic

File: src/base.jl [hunk 5]
```diff
@@ -14,18 +14,24 @@ function check_size(T, sz)
...
+function ShortString{T}(s::Union{String, SubString{String}}) where {T}
     sz = sizeof(s)
     check_size(T, sz)
     bits_to_wipe = 8(sizeof(T) - sz)
+
+    # Warning: if a SubString is at the very end of a string, which is at the end of allocated
+    # memory, this can cause an access violation, by trying to access past the end
+    # (for example, reading a 1 byte substring at the end of a length 119 string, could go past
+    # the end)
+
```
Comment (xiaodaigh) [inline src/base.jl:26]: we need a proper fix for this # #

**Review by xiaodaigh — APPROVED** [review]

**Pull request merged**